content/post/the-lab-book.md

```diff
@@ -0,0 +1,87 @@
+
+Often, the iteration that a data scientist goes through are opaque to their collaborators -- they may share a working (and reproducible) python notebook, but it’s often impossible to recover the different iterations the DS went through to get to that particular configuration, or what was learned along the way. While using Git can help, the way data scientists use git often does not capture the scientific process of hypothesis-experiment-analysis-conclusion. Instead, you end up with one commit that says “model works” and you don’t have a record of all of the different things the data scientist observed and learned along the way.
+
+When it comes to understanding *_why_* a given model is structured the way it is, you need a history of what was already tried and what worked and what didn’t. If someone on the team wants to pick up where the first practitioner left off, they can avoid lots of wasted effort if they know what was attempted before.
+
+## When to use the lab book?
+
+Have you ever found yourself in a situation where you are working on a gnarly piece of code and you feel like you are just slamming your head into a brick wall over and over and over again? This happens to me pretty frequently and, if you’re around me while I’m going through this, you’ll hear me repeatedly say, aloud, “how on earth could *_that_* be happening?”. At about the third time I say that phrase aloud, I know I need to bust out the lab book.
+
+Software complexity (especially in the data science space) can make it difficult or impossible to reason about what’s happening. For example:
+
+* Working third-party opaque libraries
+* Stochastic processes
+* “Black-box” ML models
+* Any piece of software sufficiently complex that you can’t hold all of the state in your head at a given time
+
+At any point where you are working on a task and you are repeatedly encountering unexpected or surprising results that’s a good time to use the data science lab book to begin keeping track of your hypotheses, the evidence you have learned, and the assumptions you’re making.
+
+In particular, it’s a good idea to use a lab book when:
+
+* You’re doing experimental research (e.g., model-building and model-fitting)
+* You’ve encountered anomalous patterns in results that haven’t gone away with 3 different attempts
+* You’re isolating a bug in an opaque part of the code-base
+
+By recording your experiences, you’ll not only force yourself to think more clearly and carefully about the problem at hand, but you’ll also be building a record that your colleagues can review when they either join you to help out or want to learn from your process.
+
+## What does the lab book for data science look like?
+
+A lab book should have one entry per experiment and should be ordered in reverse-chronological order (newest experiments at the top). A google-doc works great for this, especially if you’re collaborating with other team-members.
+
+Here’s an example template:
+
+**Date:** 2019-09-15  
```

Comment:
What about tracking changes in data sets? Is this in the conditions section? I'm thinking data set changes like added October's data to the data set (once October is over), not introducing whole new data sources. More data versioning

Reply:
Yeah, that'd go in "conditions". In general, when I'm doing this I'm working with a static data set and changing one thing at a time to measure cause and affect. So it might be like "added variable x to the model ... here's what happened." Or you could do something like adding a week of data at a time to a model to see which slice causes predictive accuracy to drop. It really depends on the situation!